eos-tech-support/eos-write-image

```diff
@@ -141,7 +141,7 @@ cat_image() {
     MIME_TYPE="$(file --brief --mime-type "$IMAGE")"
     if [[ "$MIME_TYPE" =~ xz$ ]]; then
         # Image is xzipped, xz has its own rate meter
-        xzcat -v "$IMAGE"
+        xzcat -T 0 -v "$IMAGE"
```

Comment:
Have you tested whether this actually has any effect? I ask because `man xzcat` has this to say:

```
       -T threads, --threads=threads
              […]

              Threaded decompression hasn't been implemented yet.
```

Reply:
I did test it (5 trials each with and without `-T0` on the 3.5.6 base image using my favourite hammer for this kind of thing, [multitime](https://github.com/ltratt/multitime)) and the answer is "no, it makes no difference; in fact, -T0 is slower, but well within the margin of error":

```
$ cat xzcat.batch 
-q xzcat eos-eos3.5-amd64-amd64.190212-145424.base.img.xz
-q xzcat -T0 eos-eos3.5-amd64-amd64.190212-145424.base.img.xz
$ src/ltratt/multitime/multitime -b xzcat.batch -n 5
===> multitime results
1: -q xzcat eos-eos3.5-amd64-amd64.190212-145424.base.img.xz
            Mean                Std.Dev.    Min         Median      Max
real        154.990+/-8.4373      4.679       148.134     155.076     162.522     
user        154.031+/-8.0981      4.491       147.470     154.158     161.221     
sys         0.919+/-0.3264      0.181       0.652       0.884       1.212       

2: -q xzcat -T0 eos-eos3.5-amd64-amd64.190212-145424.base.img.xz
            Mean                Std.Dev.    Min         Median      Max
real        158.676+/-6.9007      3.827       151.295     159.520     162.263     
user        157.522+/-6.7639      3.751       150.250     158.524     160.913     
sys         1.083+/-0.1978      0.110       0.952       1.116       1.244       
```

I did some quick searching and a tool called [pixz](https://github.com/vasi/pixz) exists. It's in Debian Buster and appears to support parallel decompression. You could consider testing whether it actually helps, particularly since this tool (and eos-installer) write the output sequentially to disk, and if so prepare some patches to add it to our distro and adjust this script and eos-installer to use it.

Reply:
You may also be interested in some research I did into using Zstandard. https://phabricator.endlessm.com/T21041

Reply:
FYI

```
$ src/ltratt/multitime/multitime -r 'rm -f eos-eos3.5-amd64-amd64.190212-145424.base.img' -n 5 pixz -k -d *.xz
===> multitime results
1: -r "rm -f eos-eos3.5-amd64-amd64.190212-145424.base.img" pixz -k -d eos-eos3.5-amd64-amd64.190212-145424.base.img.xz
            Mean                Std.Dev.    Min         Median      Max
real        64.914+/-2.2249      1.234       62.691      65.175      66.369      
user        211.222+/-1.1262      0.625       210.525     211.300     212.192     
sys         7.891+/-0.3649      0.202       7.645       7.885       8.216       
wjt@tensionsheet-redux:~$ 
```

Reply:
Thanks for looking into it but I don't have time to devote to this right now. I only created the patch because Izzy mentioned that it's slow and I thought it would be a one-liner.